lua/nvim-treesitter/indent.lua

```diff
@@ -2,52 +2,68 @@ local parsers = require "nvim-treesitter.parsers"
 local queries = require "nvim-treesitter.query"
 local tsutils = require "nvim-treesitter.ts_utils"
 
-local M = {}
+local function get_first_node_at_line(root, lnum)
+  local col = vim.fn.indent(lnum)
+  return root:descendant_for_range(lnum - 1, col, lnum - 1, col)
+end
 
--- TODO(kiyan): move this in tsutils and document it
-local function get_node_at_line(root, lnum)
-  for node in root:iter_children() do
-    local srow, _, erow = node:range()
-    if srow == lnum then
-      return node
-    end
+local function get_last_node_at_line(root, lnum)
+  local col = #vim.fn.getline(lnum) - 1
+  return root:descendant_for_range(lnum - 1, col, lnum - 1, col)
+end
 
-    if node:child_count() > 0 and srow < lnum and lnum <= erow then
-      return get_node_at_line(node, lnum)
+local function get_delimiter_node(root, anchor, delimiter)
+  local function walk_sibling(dir, row, col, matcher)
+    local node = root:descendant_for_range(row, col, row, col)
+    local pos = 1
+    while node and not matcher(node) do
+      if dir == "backward" then
+        node = node:prev_sibling()
+      elseif dir == "forward" then
+        node = node:next_sibling()
+      else
+        error()
+      end
```

Comment:
looks like this whole thing could be replaced with `node:prev_sibling()` only

Reply:
Updated.